trie/binary.go

```diff
@@ -0,0 +1,483 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package trie
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"sort"
+
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type BinaryNode interface {
+	Hash() []byte
+	hash(off int) []byte
+	Commit() error
+	//tryGet(key []byte, depth int) ([]byte, error)
+}
+
+// BinaryHashPreimage represents a tuple of a hash and its preimage
+type BinaryHashPreimage struct {
+	Key   []byte
+	Value []byte
+}
+
+type binkey []byte
+
+type storeSlot struct {
+	key   binkey
+	value []byte
+}
+
+type store []storeSlot
+
+type BinaryTrie struct {
+	root  BinaryNode
+	store store
+}
+
+// All known implementations of binaryNode
+type (
+	// branch is a node with two children ("left" and "right")
+	// It can be prefixed by bits that are common to all subtrie
+	// keys and it can also hold a value.
+	branch struct {
+		left  BinaryNode
+		right BinaryNode
+
+		key   []byte // TODO split into leaf and branch
+		value []byte
+
+		// Used to send (hash, preimage) pairs when hashing
+		CommitCh chan BinaryHashPreimage
+
+		// This is the binary equivalent of "extension nodes":
+		// binary nodes can have a prefix that is common to all
+		// subtrees.
+		prefix binkey
+	}
+
+	hashBinaryNode []byte
+
+	empty struct{}
+)
+
+var (
+	errInsertIntoHash    = errors.New("trying to insert into a hash")
+	errReadFromHash      = errors.New("trying to read from a hash")
+	errReadFromEmptyTree = errors.New("reached an empty subtree")
+	errKeyNotPresent     = errors.New("trie doesn't contain key")
+
+	// 0_32
+	zero32 = []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}
+)
+
+func newBinKey(key []byte) binkey {
+	bits := make([]byte, 8*len(key))
+	for i := range bits {
+
+		if key[i/8]&(1<<(7-i%8)) == 0 {
+			bits[i] = 0
+		} else {
+			bits[i] = 1
+		}
+	}
+
+	return binkey(bits)
+}
+func min(i, j int) int {
+	if i < j {
+		return i
+	}
+	return j
+}
+func (b binkey) commonLength(other binkey) int {
+	for i := 0; i < len(b) && i < len(other); i++ {
+		if b[i] != other[i] {
+			return i
+		}
+	}
+	return min(len(b), len(other))
+}
+func (b binkey) samePrefix(other binkey, off int) bool {
+	return bytes.Equal(b[off:off+len(other)], other[:])
+}
+
+func (s store) Len() int { return len(s) }
+func (s store) Less(i, j int) bool {
+	for b := 0; b < len(s[i].key) && b < len(s[j].key); b++ {
+		if s[i].key[b] != s[j].key[b] {
+			// if s[j].key.Bit(b) is true, then it is
+			// the greater value of the two.
+			return s[j].key[b] == 1
+		}
+	}
+
+	// Keys are equal on their common length, the shortest
+	// is the smaller one.
+	return len(s[i].key) < len(s[j].key)
```

Comment:
I think this (or something like it) is equivalent:
```suggestion
 	return bytes.Compare(s[i].key, s[j].key) == -1
```